domain/user/state/state.go

```diff
@@ -0,0 +1,488 @@
+	if err != nil {
+		return errors.Annotate(err, "getting DB access")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(addUser(ctx, tx, uuid, user, creatorUUID))
+	})
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the creator does
+// not exist that satisfies usererrors.UserCreatorUUIDNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(
+	ctx context.Context,
+	uuid user.UUID,
+	user user.User,
+	creatorUUID user.UUID,
+	passwordHash string,
+	salt []byte,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting DB access")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		return errors.Trace(setPasswordHash(ctx, tx, uuid, passwordHash, salt))
+	})
+}
+
+// AddUserWithActivationKey will add a new user to the database with the
+// provided activation key. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the users creator
+// does not exist an error that satisfies usererrors.UserCreatorNotFound
+// will be returned.
+func (st *State) AddUserWithActivationKey(
+	ctx context.Context,
+	uuid user.UUID,
+	user user.User,
+	creatorUUID user.UUID,
+	activationKey string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting DB access")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return errors.Trace(setActivationKey(ctx, tx, uuid, activationKey))
+	})
+}
+
+// GetUser will retrieve the user specified by UUID from the database.
+// If the user does not exist an error that satisfies
+// usererrors.NotFound will be returned.
+func (st *State) GetUser(ctx context.Context, uuid user.UUID) (user.User, error) {
+	db, err := st.DB()
+	if err != nil {
+		return user.User{}, errors.Annotate(err, "getting DB access")
+	}
+
+	var usr user.User
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		getUserQuery := `
+SELECT (uuid, name, display_name, created_by_uuid, created_at) AS &User.*
+FROM user
+WHERE uuid = $M.uuid
+`
+		selectGetUserStmt, err := sqlair.Prepare(getUserQuery, User{}, sqlair.M{})
+		if err != nil {
+			return errors.Annotate(err, "preparing select getUser query")
+		}
+
+		var result User
+		err = tx.Query(ctx, selectGetUserStmt, sqlair.M{"uuid": uuid.String()}).Get(&result)
+		if err != nil && errors.Is(err, sql.ErrNoRows) {
+			return errors.Annotatef(usererrors.NotFound, "%q", uuid)
+		} else if err != nil {
+			return errors.Annotatef(err, "getting user with uuid %q", uuid)
+		}
+
+		usr = user.User{
```

Comment:
You could also add a method to your local `User` that converts it to the core type.
Then that's in one place.

Reply:
thx, added method to `state/types.go`